plugin/folding_range.py

```diff
@@ -0,0 +1,138 @@
+            view_change_count = self.view.change_count()
+            # If the stored change_count matches the view's actual change count, the request has already been run for
+            # this document state (i.e. "Edit" menu was opened before) and the results are still valid - no need to send
+            # another request.
+            if self.change_count == view_change_count:
+                return False
+            self.change_count = -1
+            session = self.best_session(self.capability)
+            if session:
+                params = {'textDocument': text_document_identifier(self.view)}  # type: FoldingRangeParams
+                session.send_request_async(
+                    Request.foldingRange(params, self.view),
+                    partial(self._handle_response_async, view_change_count)
+                )
+            return False
+        return self.folding_region is not None  # Already set or unset by self.description
+
+    def _handle_response_async(self, change_count: int, response: Optional[List[FoldingRange]]) -> None:
+        self.change_count = change_count
+        self.folding_ranges = response or []
+
+    def description(
+        self, manual: bool = True, hidden: bool = False, event: Optional[dict] = None, point: Optional[int] = None
+    ) -> str:
+        if manual:
+            return "LSP: Fold"
+        # Implementation detail of Sublime Text: TextCommand.description is called *before* TextCommand.is_visible
+        self.folding_region = None
+        if self.change_count != self.view.change_count():  # Ensure that the response has already arrived
+            return "LSP <debug>"  # is_visible will return False
+        if point is not None:
+            pt = point
+        else:
+            selection = self.view.sel()
+            if len(selection) != 1 or not selection[0].empty():
+                return "LSP <debug>"  # is_visible will return False
+            pt = selection[0].b
+        for folding_range in sorted(self.folding_ranges, key=lambda r: r['startLine'], reverse=True):
+            region = range_to_region(folding_range_to_range(folding_range), self.view)
+            if region.contains(pt) and not self.view.is_folded(region):
+                # Store the relevant folding region, so that we don't need to do the same computation again in
+                # self.is_visible and self.run
+                self.folding_region = region
+                return {
+                    FoldingRangeKind.Comment: "LSP: Fold this comment",
+                    FoldingRangeKind.Imports: "LSP: Fold imports",
+                    FoldingRangeKind.Region: "LSP: Fold this region",
+                    'array': "LSP: Fold this array",  # used by LSP-json
+                    'object': "LSP: Fold this object",  # used by LSP-json
+                }.get(folding_range.get('kind', ''), "LSP: Fold")
+        return "LSP <debug>"  # is_visible will return False
```

Comment:
Noticed interesting behavior with:

```jsonc
{
    "foo": [1,
        2,
        3
    ]
}
```

Folding regions in this case are: `[{'startLine': 15, 'endLine': 16, 'kind': 'array'}, {'startLine': 1, 'endLine': 3, 'kind': 'array'}, {'startLine': 0, 'endLine': 4, 'kind': 'object'}]`

(not sure what the first region is about but we can ignore it)

With cursor anywhere within line 1, we fold the root object.
In VSCode it applies the second fold region.

While I understand that per spec it's supposed to fold from the end of the `startLine`, maybe the cursor intersection region should actually be checked against the start of the line. It seems like a smart behavior.

More usuful example would be:

```py
def fn():
    if foo:
        'xxx'
```

where it could fold the body of `if foo` when on that line but currently it folds the the whole body of the function.

Reply:
> With cursor anywhere within line 1, we fold the root object.
> In VSCode it applies the second fold region.

I think a part of this problem is that the JSON language server doesn't report the optional `startCharacter` & `endCharacter` values. If you put the caret within `"foo"`, then the builtin `fold` command (Edit > Code Folding > Fold) also folds the root object. But if you put it between the `[` and `1`, the builtin command folds the array, as expected. If the server would use the `startCharacter`, then the `lsp_fold` behavior would be the same.

Maybe it can be improved with some heuristics if `startCharacter` is missing, but I'm unsure if this would do the right thing in all cases. For example with
```json
{
    "foo": [1, ["a",
            "b",
            "c"
        ],
        2,
        3
    ]
}
```
I think there is no way to decide which of the arrays should be folded. It would probably only makes sense to fold the outermost array in that case.

For the Python example, the builtin behavior is also to fold the function body when the cursor is within the `if foo` line. And it does nothing if the caret is on the first line. This is only the indentation based folding, but it would probably still be trivial to implement to fold the "inner" block, if the caret is on the last line before a new indentation level, if this was the desired behavior. So I'm not sure whether it would really be better to implement this in `lsp_fold`, when it differs from the builtin functionality. But I'm open for discussion.

LSP-TexLab has support for `startCharacter` & `endCharacter`, but doesn't really seem to use it advantageously.

Reply:
I have a feeling that the behavior that VSCode shows makes more sense. At least for languages like python (and to JSON with some corner-case exceptions too) where it feels natural to be able to fold a method or a class while having the cursor on the `def ...` or `class ...` line.

Also, that behavior kinda matches ST if you compare it with the folding markers. ST provides folding markers next to the `def ...` or `class ...` and folds bodies without including that line itself. I feel like comparing the LSP functionality to those markers makes more sense than comparing it to the plain `Fold` command since both the syntax-based folding and the LSP folding is supposed to be "smarter".

Reply:
Both ways have advantages and disadvantages.

I added a `strict` parameter to make this configurable; `strict = true` means ST built-in behavior (caret must be contained within the folding region) and `strict = false` means to fold the next block/range even if the caret is anywhere on the starting line (starting with innermost/rightmost range if there are multiple ranges on the same line). Using `true` as the default value, because I think an intention for the LSP commands would be to use them as a direct replacement in key bindings for the built-in ST command, so they should probably follow ST behavior as close as possible/useful. And `strict = true` allows more precise behavior when there are multiple folding ranges on a single line, dependent on the exact caret position.

> Also, that behavior kinda matches ST if you compare it with the folding markers. ST provides folding markers next to the `def ...` or `class ...` and folds bodies without including that line itself.

I think folding buttons must be considered independent of the behavior what happens when you fold via key binding or from the menu. The fold buttons in the gutter are on the line of the `def ...` because this line is where the folded region begins; in particular the linebreak (`\n`) is the first character of the folded region, and this is the place where ST places the (typically) yellow fold marker `(...)`. You can also confirm by placing the caret at the end of such a line; in that case ST built-in `fold` command indeed folds the following block. So you cannot say that the one way or the other would match the fold button placement any more or less.

When/if ST starts to allow plugins to controll the fold buttons in the gutter, we would of course also use the line from the starting position of the folding range.